R/deploy-site.R

```diff
@@ -0,0 +1,84 @@
+#' Deploy a pkgbown site on Travis-CI using Github Pages
+#'
+#' @inheritParams build_site
+#' @param tarball The location of the built package tarball. The default Travis
+#'   configuration for R packages sets `PKG_TARBALL` to this path.
+#' @param ssh_id The private id to use, a base64 encoded value, by default this
+#'   is read from the id_rsa environment variable. It is recommended you add this
+#'   key as a private environment variable in the Travis repository settings.
+#' @param ssh_id_file The file location to write the ssh private key obtained from `ssh_id`.
+#' @param repo_slug The `user/repo` slug for the repository.
+#' @param commit_message The commit message to be used for the commit.
+#' @param verbose Print verbose output
+#' @param ... Additional arguments passed to [build_site()].
+#' @export
+deploy_site <- function(
+  pkg = ".",
+  tarball = Sys.getenv("PKG_TARBALL", ""),
+  ssh_id = Sys.getenv("id_rsa", ""),
+  ssh_id_file = "~/.ssh/id_rsa",
+  repo_slug = Sys.getenv("TRAVIS_REPO_SLUG", ""),
+  commit_message = "build site",
+  verbose = FALSE,
+  ...) {
+
+  if (!nzchar(tarball)) {
+    stop("No built tarball detected, please provide the location of one with `tarball`", call. = FALSE)
+  }
+  rule("Deploying site", line = 2)
+  rule("Installing package", line = 1)
+  callr::rcmd("INSTALL", tarball, show = verbose, fail_on_status = TRUE)
+
+  build_site(pkg, ...)
+
+  if (!nzchar(ssh_id)) {
+    stop("Need to setup a deploy key with `ssh-keygen`, base64 encode it and then assign it to `id_rsa`", call. = FALSE)
+  }
+
+  if (!nzchar(repo_slug)) {
+    stop("No repo detected, please supply one with `repo_slug`", call. = FALSE)
+  }
+
+  remote_url <- glue::glue("git@github.com:{repo_slug}.git")
+
+
+  rule("Setting up SSH id", line = 1)
+  cat_line("Copying private key to: ", ssh_id_file)
+  write_lines(rawToChar(openssl::base64_decode(ssh_id)), ssh_id_file)
+  cat_line("Setting private key permissions", ssh_id_file)
+  fs::file_chmod(ssh_id_file, "0600")
+
+
+  rule("Commiting built site", line = 1)
+  withr::with_dir("docs", {
+    processx::run("git", "init", echo = verbose)
+    processx::run("git", c("add", "-A", "."), echo = verbose)
+    processx::run("git", c("remote", "add", "origin", remote_url), echo = verbose)
+    processx::run("git", c("commit", "-m", commit_message), echo = verbose)
+
+    rule("Deploying to GitHub Pages", line = 1)
+    processx::run("git", c("remote", "-v"), echo = verbose)
+    processx::run("git", c("push", "--force", "origin", "HEAD:gh-pages"), echo = verbose)
+  })
+
+  rule("Deploy completed", line = 2)
+}
+
+#' Generate a new rsa private key and encode it
+#'
+#' This key should be added to the [encrypted environment
+#' variables](https://docs.travis-ci.com/user/environment-variables/#defining-variables-in-repository-settings)
+#' in your travis settings.
+#'
+#' @inheritParams openssl::rsa_keygen
+#' @export
+deploy_keygen <- function(bits = 4096) {
```

Comment:
It is possible this should live in usethis instead?

I need to think about a good API for this, you need both the public key (so you can add it as a deploy key on GitHub) as well as the base64 encoded private key (that you need to put in your Travis encrypted variables).

Reply:
Yeah, usethis might be a better home - could be `use_travis_key()` and would display advice about how to set in travis/github

Reply:
Realized now that this is `travis::use_travis_deploy()`.